store/postgres/src/relational_queries.rs

```diff
@@ -223,6 +224,126 @@ impl<'a> QueryFragment<Pg> for QueryValue<'a> {
     }
 }
 
+#[derive(Copy, Clone, PartialEq)]
+enum Comparison {
+    Less,
+    LessOrEqual,
+    Equal,
+    NotEqual,
+    GreaterOrEqual,
+    Greater,
+}
+
+impl Comparison {
+    fn as_str(&self) -> &str {
+        use Comparison::*;
+        match self {
+            Less => " < ",
+            LessOrEqual => " <= ",
+            Equal => " = ",
+            NotEqual => " != ",
+            GreaterOrEqual => " >= ",
+            Greater => " > ",
+        }
+    }
+}
+
+/// Produce a comparison between the string column `column` and the string
+/// value `text` that makes it obvious to Postgres' optimizer that it can
+/// first consult the partial index on `left(column, 2048)` instead of going
+/// straight to a sequential scan of the underlying table. We do this by
+/// writing the comparison `column op text` in a way that
+/// involves `left(column, 2048)`
+#[derive(Constructor)]
+struct PrefixComparison<'a> {
+    op: Comparison,
+    column: &'a Column,
+    text: &'a Value,
+}
+
+impl<'a> PrefixComparison<'a> {
+    fn push_column_prefix(&self, mut out: AstPass<Pg>) -> QueryResult<()> {
+        out.push_sql("left(");
+        out.push_identifier(self.column.name.as_str())?;
+        out.push_sql(", ");
+        out.push_sql(&STRING_PREFIX_SIZE.to_string());
+        out.push_sql(")");
+        Ok(())
+    }
+
+    fn push_value_prefix(&self, mut out: AstPass<Pg>) -> QueryResult<()> {
+        out.push_sql("left(");
+        QueryValue(self.text, &self.column.column_type).walk_ast(out.reborrow())?;
+        out.push_sql(", ");
+        out.push_sql(&STRING_PREFIX_SIZE.to_string());
+        out.push_sql(")");
+        Ok(())
+    }
+
+    fn push_prefix_cmp(&self, op: Comparison, mut out: AstPass<Pg>) -> QueryResult<()> {
+        self.push_column_prefix(out.reborrow())?;
+        out.push_sql(op.as_str());
+        self.push_value_prefix(out.reborrow())
+    }
+
+    fn push_full_cmp(&self, op: Comparison, mut out: AstPass<Pg>) -> QueryResult<()> {
+        out.push_identifier(self.column.name.as_str())?;
+        out.push_sql(op.as_str());
+        QueryValue(self.text, &self.column.column_type).walk_ast(out)
+    }
+}
+
+impl<'a> QueryFragment<Pg> for PrefixComparison<'a> {
+    fn walk_ast(&self, mut out: AstPass<Pg>) -> QueryResult<()> {
+        use Comparison::*;
+
+        // Write uv for a string where u is a string at most N long and v is
+        // empty if u is shorter than N, same for another string st.
+        // For the comparison operators '=', '!=', '<', '>' we have the
+        // relation
+        //   uv op st <=> u op s || u = s && uv op st
+        //            <=> u op s || u = s &&  v op  t
+        // If t is the empty string, v op t will always be false so that the
```

Comment:
> If t is the empty string, v op t will always be false

This is not true since v is any string. The case where it's sufficient to compare the prefix is not so simple to explain, I don't know what would be the best way.

Reply:
Yeah, I've been struggling with writing this up cleanly. What do you think of the following, which would require a small code change, and I agree there need to be more tests:

Looking at the comparison operators besides `<=` and `>=`, the general statement
```
      uv op st <=> u op s || u = s && uv op st
```
is true because the comparisons are all lexicographic, i.e., there's a first position in the two strings where the characters at that position compare according to `op`, and up to that position the two strings are equal.

We want to know under what circumstances we do not need to check `u = s && uv op st`. If we look at strings `st` where `len(st) <= N - 1`, `t` will be empty, and `len(s) <= N - 1`.

Since we compare prefixes of length `N`, either `len(uv) <= N - 1`, in which case `v` is empty, and `u op s = uv op st`, or `len(uv) >= N`, and therefore `len(u) = N`, `u != s` so that the check `u = s && uv op st` will always be false.

Reply:
And for `<=` and `>=` we apply the above logic to (using `<=` as an example`)
```
  uv <= st <=> u < s || u = s && uv <= st
```

Reply:
For `=` and `!=` this boils down to
```
  uv = st <=> u = s || u = s && uv = st <=> u = s && uv = st
  uv != st <=> u != s || u = s && uv != st <=> u != s || uv != st
```

The `!=` case won't buy us much, since the `|| uv != st` will make Postgres not use the index in almost all cases.

Reply:
This is mostly making sense to me, but `uv = st <=> u = s || ...` is not true.

Reply:
Yeah, you're right. I'll have another attempt at cleaning this up.